analyzers/src/SonarAnalyzer.CSharp/Rules/ArgumentSpecifiedForCallerInfoParameter.cs

```diff
@@ -34,55 +34,27 @@ public sealed class ArgumentSpecifiedForCallerInfoParameter : SonarDiagnosticAna
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+        private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
 
         private static readonly ImmutableArray<KnownType> CallerInfoAttributesToReportOn =
             ImmutableArray.Create(
+                KnownType.System_Runtime_CompilerServices_CallerArgumentExpressionAttribute,
                 KnownType.System_Runtime_CompilerServices_CallerFilePathAttribute,
-                KnownType.System_Runtime_CompilerServices_CallerLineNumberAttribute
-            );
+                KnownType.System_Runtime_CompilerServices_CallerLineNumberAttribute);
 
-        protected override void Initialize(SonarAnalysisContext context)
-        {
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                if (new CSharpMethodParameterLookup((InvocationExpressionSyntax)c.Node, c.SemanticModel) is { MethodSymbol: { } } methodParameterLookup
+                    && methodParameterLookup.GetAllArgumentParameterMappings() is { } argumentMappings)
                 {
-                    var methodCall = (InvocationExpressionSyntax)c.Node;
-                    var methodParameterLookup = new CSharpMethodParameterLookup(methodCall, c.SemanticModel);
-
-                    var methodSymbol = methodParameterLookup.MethodSymbol;
-                    if (methodSymbol == null)
+                    foreach (var argumentMapping in argumentMappings.Where(x => x.Symbol.GetAttributes(CallerInfoAttributesToReportOn).Any()))
                     {
-                        return;
+                        c.ReportIssue(Diagnostic.Create(Rule, argumentMapping.Node.GetLocation()));
                     }
-
-                    var argumentMappings = methodParameterLookup.GetAllArgumentParameterMappings();
-                    foreach (var argumentMapping in argumentMappings)
-                    {
-                        var parameter = argumentMapping.Symbol;
-                        var argument = argumentMapping.Node;
-
-                        var callerInfoAttributeDataOnCall = GetCallerInfoAttribute(parameter);
-                        if (callerInfoAttributeDataOnCall == null)
-                        {
-                            continue;
-                        }
-
-                        if (c.SemanticModel.GetSymbolInfo(argument.Expression).Symbol is IParameterSymbol symbolForArgument &&
-                            Equals(callerInfoAttributeDataOnCall.AttributeClass, GetCallerInfoAttribute(symbolForArgument)?.AttributeClass))
```

Comment:
The 
`Equals(callerInfoAttributeDataOnCall.AttributeClass, GetCallerInfoAttribute(symbolForArgument)?.AttributeClass)` part doesn't make sense IMO. The same test was done in line 67 before.

Reply:
The `c.SemanticModel.GetSymbolInfo(argument.Expression).Symbol is IParameterSymbol symbolForArgument` is also useless. The `x.Symbol` in argumentMappings refer to this parameter already.

Reply:
Nice rewrite.